ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java

```diff
@@ -867,8 +867,25 @@ public boolean delistResource(XAResource xaRes, int flags)
 					{
 						if (XAUtils.canOptimizeDelist(xaRes))
 						{
-							xaRes.end(info.xid(), XAResource.TMFAIL);
-							xaRes.rollback(info.xid());
+							boolean rollbackOnly = false;
+							try {
+								xaRes.end(info.xid(), XAResource.TMFAIL);
+							} catch (XAException e1) {
+								if ((e1.errorCode >= XAException.XA_RBBASE)
+										&& (e1.errorCode < XAException.XA_RBEND)) {
+									rollbackOnly = true;
```

Comment:
I would prefer to match the similar handling here: https://github.com/jbosstm/narayana/blob/6810af48583e5cf889310dbd9b83986527b074a3/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java#L278-L279 as you say they are the same values.

Specifically XA_RBROLLBACK is defined as XA_RBBASE and XA_RBEND is defined as XA_RBTRANSIENT

Reply:
Fair.
What about the `<=` ? Is there any particular reason why XA_RBEND is excluded here?

Reply:
Because I included it from the other place I referenced :)

Definitely though for this case I think it's OK to have <= as XA_RBTRANSIENT is still declared as marking it as rollback only for XAResource::end. I will update it here and the other place. Thanks for noticing that!

Reply:
I have updated it to include this, and BTW while I was testing it (I changed the XA_RB to the transient one). I noticed that the boolean logic for checking what to throw or not was not right and thus have updated that too.

(At least for now) you can see my changes in https://github.com/jbosstm/narayana/compare/1aa8eb9436a8c385f34640261cf33735643a72dd..38a096e87efc1b41108ef2d9e0a26b702079733c